File: amplify/functions/shared/email-mime-intake.ts
import {
  directCitationRationale,
  extractDirectCitations,
  titleFromUrl,
} from "./email-submission";

const HREF_PATTERN = /href\s*=\s*(?:"([^"]+)"|'([^']+)'|([^\s>]+))/gi;
const SKIP_URL_MARKERS = [
  "unsubscribe",
  "list-unsubscribe",
  "/unsubscribe",
  "mailchi.mp/unsubscribe",
  "/opt-out",
  "preferences",
  "email-preferences",
  "doubleclick.net",
  "facebook.com/sharer",
  "twitter.com/intent",
  "linkedin.com/sharing",
  "fonts.googleapis.com",
];
const SKIP_URL_PREFIXES = ["mailto:", "tel:", "javascript:", "cid:", "#"];

function decodeHtmlEntities(value: string): string {
  return value
    .replace(/&amp;/gi, "&")
    .replace(/&lt;/gi, "<")
    .replace(/&gt;/gi, ">")
    .replace(/&quot;/gi, '"')
    .replace(/&#39;/g, "'");
}

export function normalizeHrefUrl(raw: string): string | null {
  let url = decodeHtmlEntities(String(raw ?? "").trim());
  if (!url) return null;
  if (url.startsWith("//")) url = `https:${url}`;
  if (url.startsWith("<") && url.endsWith(">")) url = url.slice(1, -1).trim();
  const lowered = url.toLowerCase();
  if (SKIP_URL_PREFIXES.some((prefix) => lowered.startsWith(prefix))) return null;
  if (!/^https?:\/\//i.test(url)) return null;
  return url.replace(/[.,);]+$/, "");
}

export function isSkippableNewsletterUrl(url: string): boolean {
  const lowered = String(url ?? "").toLowerCase();
  if (SKIP_URL_MARKERS.some((marker) => lowered.includes(marker))) return true;
  if (/\.(png|jpe?g|gif|webp)(\?|$)/i.test(lowered)) return true;
  return false;
}

export function extractHrefUrlsFromHtml(html: string): string[] {
  const urls: string[] = [];
  const seen = new Set<string>();
  for (const match of html.matchAll(HREF_PATTERN)) {
    const candidate = normalizeHrefUrl(match[1] ?? match[2] ?? match[3] ?? "");
    if (!candidate || seen.has(candidate) || isSkippableNewsletterUrl(candidate)) continue;
    seen.add(candidate);
    urls.push(candidate);
  }
  return urls;
}

export function extractDirectCitationsFromIntake(input: {
  bodyText: string;
  htmlParts?: string[];
}): Array<{ kind: string; url: string; title: string; ingestion_rationale: string; doi?: string }> {
  const citations = extractDirectCitations(input.bodyText);
  const seen = new Set(citations.map((entry) => entry.url));
  for (const htmlPart of input.htmlParts ?? []) {
    for (const url of extractHrefUrlsFromHtml(htmlPart)) {
      if (seen.has(url)) continue;
      seen.add(url);
      citations.push({
        kind: "url",
        url,
        title: titleFromUrl(url),
        ingestion_rationale: directCitationRationale(url),
      });
    }
  }
  return citations;
}

function decodeQuotedPrintable(value: string): string {
  return value
    .replace(/=\r?\n/g, "")
    .replace(/=([0-9A-F]{2})/gi, (_, hex: string) => String.fromCharCode(parseInt(hex, 16)));
}

function decodePartBody(body: string, transferEncoding: string): string {
  const encoding = transferEncoding.trim().toLowerCase();
  if (encoding === "base64") {
    return Buffer.from(body.replace(/\s/g, ""), "base64").toString("utf8");
  }
  if (encoding === "quoted-printable") {
    return decodeQuotedPrintable(body);
  }
  return body;
}

function htmlToText(html: string): string {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<[^>]+>/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function collectMimeBodies(raw: string): { plain: string[]; html: string[]; subject: string } {
  const subjectMatch = raw.match(/^Subject:\s*(.*)$/im);
  const subject = subjectMatch?.[1]?.trim() ?? "";
  const plain: string[] = [];
  const html: string[] = [];
  const partPattern =
    /Content-Type:\s*(text\/(?:plain|html))[^\r\n]*(?:[\r\n]+(?![\r\n])[^\r\n]+)*[\r\n]+(?:Content-Transfer-Encoding:\s*([^\r\n]+)[\r\n]+)?[\r\n]+([\s\S]*?)(?=[\r\n]--[^\r\n]+[\r\n]|$)/gi;
  let match: RegExpExecArray | null = partPattern.exec(raw);
  while (match) {
    const contentType = match[1]?.toLowerCase() ?? "";
    const encoding = match[2] ?? "";
    const body = decodePartBody(match[3] ?? "", encoding);
    if (contentType === "text/plain" && body.trim()) plain.push(body.trim());
    if (contentType === "text/html" && body.trim()) html.push(body.trim());
    match = partPattern.exec(raw);
  }
  return { plain, html, subject };
}

function preprocessRawMimeForLinkScan(raw: string): string {
  return raw
    .replace(/=\r?\n/g, "")
    .replace(/=([0-9A-F]{2})/gi, (_, hex: string) => String.fromCharCode(parseInt(hex, 16)));
}

function extractHrefUrlsFromRawMime(raw: string): string[] {
  const decoded = preprocessRawMimeForLinkScan(raw);
  const urls: string[] = [];
  const seen = new Set<string>();
  for (const match of decoded.matchAll(HREF_PATTERN)) {
    const candidate = normalizeHrefUrl(match[1] ?? match[2] ?? match[3] ?? "");
    if (!candidate || seen.has(candidate) || isSkippableNewsletterUrl(candidate)) continue;
    seen.add(candidate);
    urls.push(candidate);
  }
  return urls;
}

export function parseInboundMimeForIntake(rawBytes: Uint8Array): {
  subject: string;
  bodyText: string;
  htmlParts: string[];
  citations: Array<{ kind: string; url: string; title: string; ingestion_rationale: string; doi?: string }>;
} {
  const raw = Buffer.from(rawBytes).toString("utf8");
  const { plain, html, subject } = collectMimeBodies(raw);
  let bodyText = plain.join("\n\n").trim();
  if (!bodyText && html.length > 0) {
    bodyText = htmlToText(html.join("\n\n"));
  }
  const citations = extractDirectCitationsFromIntake({ bodyText, htmlParts: html });
  if (citations.length === 0) {
    const seen = new Set(citations.map((entry) => entry.url));
    for (const url of extractHrefUrlsFromRawMime(raw)) {
      if (seen.has(url)) continue;
      seen.add(url);
      citations.push({
        kind: "url",
        url,
        title: titleFromUrl(url),
        ingestion_rationale: directCitationRationale(url),
      });
    }
  }
  return {
    subject: subject || "(no subject)",
    bodyText,
    htmlParts: html,
    citations,
  };
}

File: src/papyrus_newsroom/email_mime_intake.py
"""MIME parsing helpers for inbound email citation extraction."""

from __future__ import annotations

import html as html_module
import re
from email import policy
from email.parser import BytesParser
from typing import Any
from urllib.parse import urlparse

from papyrus_newsroom.email_submissions import (
    _DOI_PATTERN,
    _URL_PATTERN,
    _direct_citation_rationale,
    _html_to_text,
    _title_from_url,
)

_HREF_PATTERN = re.compile(
    r"""href\s*=\s*(?:"([^"]+)"|'([^']+)'|([^\s>]+))""",
    re.IGNORECASE,
)
_SKIP_URL_MARKERS = (
    "unsubscribe",
    "list-unsubscribe",
    "/unsubscribe",
    "mailchi.mp/unsubscribe",
    "/opt-out",
    "preferences",
    "email-preferences",
    "doubleclick.net",
    "facebook.com/sharer",
    "twitter.com/intent",
    "linkedin.com/sharing",
    "fonts.googleapis.com",
    "schema.org",
    "w3.org/1999",
)
_SKIP_URL_PREFIXES = ("mailto:", "tel:", "javascript:", "cid:", "#")


def _normalize_href_url(raw: str) -> str | None:
    url = html_module.unescape(str(raw or "").strip())
    if not url:
        return None
    if url.startswith("//"):
        url = f"https:{url}"
    if url.startswith("<") and url.endswith(">"):
        url = url[1:-1].strip()
    lowered = url.lower()
    if any(lowered.startswith(prefix) for prefix in _SKIP_URL_PREFIXES):
        return None
    if not re.match(r"^https?://", url, re.IGNORECASE):
        return None
    return url.rstrip(".,);]")


def is_skippable_newsletter_url(url: str) -> bool:
    lowered = str(url or "").lower()
    if any(marker in lowered for marker in _SKIP_URL_MARKERS):
        return True
    host = (urlparse(lowered).netloc or "").lower()
    if host.endswith((".png", ".jpg", ".jpeg", ".gif", ".webp")):
        return True
    return False


def extract_href_urls_from_html(html: str) -> list[str]:
    urls: list[str] = []
    seen: set[str] = set()
    for match in _HREF_PATTERN.finditer(str(html or "")):
        candidate = _normalize_href_url(match.group(1) or match.group(2) or match.group(3) or "")
        if not candidate or candidate in seen:
            continue
        if is_skippable_newsletter_url(candidate):
            continue
        seen.add(candidate)
        urls.append(candidate)
    return urls


def _citation_records_from_urls(urls: list[str]) -> list[dict[str, str]]:
    citations: list[dict[str, str]] = []
    seen: set[str] = set()
    for url in urls:
        if url in seen:
            continue
        seen.add(url)
        citations.append(
            {
                "kind": "url",
                "url": url,
                "title": _title_from_url(url),
                "ingestion_rationale": _direct_citation_rationale(url),
            }
        )
    return citations


def extract_direct_citations_from_intake_text(*, body_text: str, html_parts: list[str] | None = None) -> list[dict[str, str]]:
    """Collect citations from visible text and from HTML anchor hrefs."""
    urls: list[str] = []
    seen: set[str] = set()

    def add_url(url: str | None) -> None:
        if not url or url in seen or is_skippable_newsletter_url(url):
            return
        seen.add(url)
        urls.append(url)

    for match in _URL_PATTERN.finditer(str(body_text or "")):
        add_url(match.group(0).rstrip(".,);]"))

    for html in html_parts or []:
        for href in extract_href_urls_from_html(html):
            add_url(href)

    citations = _citation_records_from_urls(urls)

    for match in _DOI_PATTERN.finditer(str(body_text or "")):
        doi = match.group(0).rstrip(".,);]")
        doi_url = f"https://doi.org/{doi}"
        if doi_url in seen:
            continue
        seen.add(doi_url)
        citations.append(
            {
                "kind": "doi",
                "url": doi_url,
                "doi": doi,
                "title": f"DOI {doi}",
                "ingestion_rationale": _direct_citation_rationale(doi_url),
            }
        )
    return citations


def collect_mime_text_parts(raw_bytes: bytes) -> tuple[str, list[str], list[str]]:
    """Return subject, plain text bodies, and raw HTML bodies from a MIME message."""
    message = BytesParser(policy=policy.default).parsebytes(raw_bytes)
    plain_parts: list[str] = []
    html_parts: list[str] = []
    if message.is_multipart():
        for part in message.walk():
            if part.is_multipart():
                continue
            disposition = str(part.get_content_disposition() or "").lower()
            if disposition == "attachment":
                continue
            content_type = str(part.get_content_type() or "").lower()
            try:
                payload = part.get_content()
            except Exception:
                payload = part.get_payload(decode=True)
            if content_type == "text/plain" and isinstance(payload, str):
                plain_parts.append(payload)
            elif content_type == "text/html":
                if isinstance(payload, str):
                    html_parts.append(payload)
                elif isinstance(payload, (bytes, bytearray)):
                    html_parts.append(
                        bytes(payload).decode(part.get_content_charset() or "utf-8", errors="replace")
                    )
    else:
        try:
            payload = message.get_content()
        except Exception:
            payload = message.get_payload(decode=True)
        content_type = str(message.get_content_type() or "").lower()
        if isinstance(payload, str):
            if content_type == "text/html":
                html_parts.append(payload)
            else:
                plain_parts.append(payload)
        elif isinstance(payload, (bytes, bytearray)):
            decoded = bytes(payload).decode(message.get_content_charset() or "utf-8", errors="replace")
            if content_type == "text/html":
                html_parts.append(decoded)
            else:
                plain_parts.append(decoded)
    subject = str(message.get("subject") or "").strip()
    body_text = "\n\n".join(part.strip() for part in plain_parts if part.strip())
    if not body_text and html_parts:
        body_text = _html_to_text("\n\n".join(html_parts))
    return subject, plain_parts, html_parts


def _preprocess_raw_mime_for_link_scan(raw: str) -> str:
    without_soft_breaks = re.sub(r"=\r?\n", "", raw)
    return re.sub(
        r"=([0-9A-F]{2})",
        lambda match: chr(int(match.group(1), 16)),
        without_soft_breaks,
        flags=re.IGNORECASE,
    )


def _extract_href_urls_from_raw_mime(raw_bytes: bytes) -> list[str]:
    raw = _preprocess_raw_mime_for_link_scan(raw_bytes.decode("utf-8", errors="replace"))
    urls: list[str] = []
    seen: set[str] = set()
    for match in _HREF_PATTERN.finditer(raw):
        candidate = _normalize_href_url(match.group(1) or match.group(2) or match.group(3) or "")
        if not candidate or candidate in seen or is_skippable_newsletter_url(candidate):
            continue
        seen.add(candidate)
        urls.append(candidate)
    return urls


def parse_inbound_mime_for_intake(raw_bytes: bytes) -> dict[str, Any]:
    subject, _plain_parts, html_parts = collect_mime_text_parts(raw_bytes)
    body_text = "\n\n".join(part.strip() for part in _plain_parts if part.strip())
    if not body_text and html_parts:
        body_text = _html_to_text("\n\n".join(html_parts))
    citations = extract_direct_citations_from_intake_text(body_text=body_text, html_parts=html_parts)
    if not citations:
        fallback_urls = _extract_href_urls_from_raw_mime(raw_bytes)
        if fallback_urls:
            citations = _citation_records_from_urls(fallback_urls)
    return {
        "subject": subject,
        "bodyText": body_text.strip(),
        "htmlParts": html_parts,
        "citations": citations,
    }
